packages/angular_devkit/build_angular/src/angular-cli-files/plugins/webpack-rollup-loader.ts

```diff
@@ -0,0 +1,148 @@
+import webpack = require('webpack');
+
+function splitRequest(request: string) {
+  const inx = request.lastIndexOf('!');
+  if (inx === -1) {
+    return {
+      loaders: '',
+      resource: request,
+    };
+  } else {
+    return {
+      loaders: request.slice(0, inx + 1),
+      resource: request.slice(inx + 1),
+    };
+  }
+}
+
+// Load resolve paths using Webpack.
+function webpackResolutionPlugin(
+  loaderContext: webpack.loader.LoaderContext,
+  entryId: string,
+  entryIdCodeAndMap: { code: string, map: RawSourceMap },
+) {
+  return {
+    name: 'webpack-resolution-plugin',
+    resolveId: (id: string, importerId: string) => {
+      if (id === entryId) {
+        return entryId;
+      } else {
+        return new Promise((resolve, reject) => {
+          // split apart resource paths because Webpack's this.resolve() can't handle `loader!`
+          // prefixes
+          const parts = splitRequest(id);
+          const importerParts = splitRequest(importerId);
+
+          // resolve the full path of the imported file with Webpack's module loader
+          // this will figure out node_modules imports, Webpack aliases, etc.
+          loaderContext.resolve(
+            dirname(importerParts.resource),
+            parts.resource,
+            (err, fullPath) => err ? reject(err) : resolve(parts.loaders + fullPath),
+          );
+        });
+      }
+    },
+    load: (id: string) => {
+      if (id === entryId) {
+        return entryIdCodeAndMap;
+      }
+
+      return new Promise((resolve, reject) => {
+        // load the module with Webpack
+        // this will apply all relevant loaders, etc.
+        loaderContext.loadModule(
+          id,
+          (err, source, map) => err ? reject(err) : resolve({ code: source, map: map }),
+        );
+      });
+    },
+  };
+}
+
+export default function webpackRollupLoader(
+  this: webpack.loader.LoaderContext,
+  source: string,
+  sourceMap: RawSourceMap,
+) {
+  // Note: this loader isn't cacheable because it will add the lazy chunks to the
+  // virtual file system on completion.
+  const callback = this.async();
+  if (!callback) {
+    throw new Error('Async loader support is required.');
+  }
+  const options = this.query || {};
+  const entryId = this.resourcePath;
+  const sourcemap = this.sourceMap;
+
+  // Get the VirtualFileSystemDecorator that AngularCompilerPlugin added so we can write to it.
+  // Since we use webpackRollupLoader as a post loader, this should be there.
+  // TODO: we should be able to do this in a more elegant way by again decorating webpacks
+  // input file system inside a custom WebpackRollupPlugin, modelled after AngularCompilerPlugin.
+  const vfs = this._compiler.inputFileSystem as VirtualFileSystemDecorator;
+  const virtualWrite = (path: string, data: string) =>
+    vfs.getWebpackCompilerHost().writeFile(path, data, false);
+
+  // Bundle with Rollup
+  const rollupOptions = {
+    ...options,
+    input: entryId,
```

Comment:
Question: should we disable treeshaking, such as pure functional calls removals if scripts optimisation is false?

Reply:
I think we don't even want to use this if script optimization if false. But I haven't gotten around to further matching options into the rollup pass. I think it makes more sense to do that later after we start getting feedback from its shortcomings. We might have to add a lot of config then.

Reply:
I think we don't even want to use this if script optimization if false. But I haven't gotten around to further matching options into the rollup pass. I think it makes more sense to do that later after we start getting feedback from its shortcomings. We might have to add a lot of config then.